Partial implemention of new Releases API

mozbhearsum opened the pull request:
This is built on top of #1199 and implements part of the new Releases API, specifically:
* Getting Release info (both existing and scheduled changes)
* Creating or modifying Releases (both directly and with a scheduled change)
* Critical underlying code to support these and future API endpoints -- release splitting logic, error handlers, etc.

Unlike the existing API, both direct modifications and scheduled changes are handled by the same endpoint. Passing `when` will put you into "schedule a change" mode. Notably, this also means the response has two different formats -- one that returns just new data versions, and one that includes information about the changes that were scheduled. I _think_ this makes for a much cleaner API overall, but I'm open to other opinions.

nthomas-mozilla commented:
For easier [diffing](https://github.com/mozbhearsum/balrog/compare/1-add-new-releases-tables...mozbhearsum:2-new-releases-api) could you please rebase your `1-add-new-releases-tables` branch so it's on the same base rev as this one (7a9d6e3). Thanks.

mozbhearsum commented:
> For easier [diffing](https://github.com/mozbhearsum/balrog/compare/1-add-new-releases-tables...mozbhearsum:2-new-releases-api) could you please rebase your `1-add-new-releases-tables` branch so it's on the same base rev as this one ([7a9d6e3](https://github.com/mozilla-releng/balrog/commit/7a9d6e3f53723e33e354e40b3ce0275bb956c5b4)). Thanks.

Ack, sorry! Will do.

mozbhearsum commented:
This should be up to date now.

mozbhearsum commented on src/auslib/web/admin/swagger/api_v2.yml:
```diff
@@ -0,0 +1,178 @@
+                                type: number
+                              complete:
+                                type: boolean
+  /releases/{name}:
```
I think those suggestions make sense; I'll make those changes.

> Will the get method for /releases/{name} come in a later PR ?

Yep!

mozbhearsum commented on src/auslib/web/admin/swagger/api_v2.yml:
```diff
@@ -0,0 +1,178 @@
+                        read_only:
+                          type: boolean
+                          nullable: true
+                        rule_info:
```
This is largely because I want to see what the full spec looks like before trying to refactor it. In the old spec, we ended up with a lot of very similar but different entries for the same thing[1]. The next PR will complete the API, and I'll do a refactoring pass on the schema.

[1] Eg: product has 5 definitions: https://github.com/mozilla-releng/balrog/blob/master/src/auslib/web/common/swagger/definitions.yml#L87, https://github.com/mozilla-releng/balrog/blob/master/src/auslib/web/common/swagger/parameters.yml#L2, https://github.com/mozilla-releng/balrog/blob/master/src/auslib/web/common/swagger/parameters.yml#L11, https://github.com/mozilla-releng/balrog/blob/master/src/auslib/web/common/swagger/definitions.yml#L237, https://github.com/mozilla-releng/balrog/blob/master/src/auslib/web/admin/swagger/api.yml#L131

nthomas-mozilla replied:
Makes sense, thanks.

mozbhearsum commented on tests/data/Firefox-56.0-build1.json:
```diff
@@ -0,0 +1 @@
+{"appVersion": "56.0", "detailsUrl": "https://www.mozilla.org/%LOCALE%/firefox/56.0/releasenotes/", "displayVersion": "56.0", "fileUrls": {"*": {"completes": {"*": "http://download.mozilla.org/?product=firefox-56.0-complete&os=%OS_BOUNCER%&lang=%LOCALE%"}, "partials": {"Firefox-54.0.1-build1": "http://download.mozilla.org/?product=firefox-56.0-partial-54.0.1&os=%OS_BOUNCER%&lang=%LOCALE%", "Firef ...
```
Yeah, they're straight imports with some locale removals to speed things up (the tests are largely O(locales), so it makes a big difference).

nthomas-mozilla replied:
👍

mozbhearsum commented on tests/admin/views/test_releases_v2.py:
```diff
@@ -0,0 +1,1173 @@
+    new_data_versions["."] = 1
+    for pname, pdata in firefox_62_0_build1["platforms"].items():
+        for lname in pdata.get("locales", {}):
+            new_data_versions["platforms"][pname]["locales"][lname] = 1
```
Huh, this is neat! And probably much safer/simpler.

mozbhearsum commented on tests/admin/views/test_releases_v2.py:
```diff
@@ -0,0 +1,1173 @@
+
+
+@pytest.mark.usefixtures("releases_db", "mock_verified_userinfo")
+def test_overwrite_succeeds_for_new_release(api, firefox_62_0_build1):
```
I can see why it's somewhat confusing. Overwrite doesn't obviously mean "replace existing object or create it if it doesn't exist".

mozbhearsum replied:
I think s/overwrite/insert/ might be confusing in a different way, since the endpoint and tests handle updates and deletes too. What do you think about s/overwrite/set/?

nthomas-mozilla replied:
Sounds good.

mozbhearsum commented on src/auslib/services/releases.py:
```diff
@@ -0,0 +1,379 @@
+release_merger = Merger(
+    [
+        # Merge dictionaries
+        (dict, ["merge"])
```
That's a good point. Nightlies are intended to be in scope here, so we obviously need to get that right. I'll add some tests for this, and otherwise make sure it won't be an issue.

mozbhearsum replied:
After some consideration and discussion, I think the right thing to do is to have POSTs overwrite `completes` and `partials` (if they are passed). We don't have a known use case for wanting to append to them, and the majority of the people I asked thought overwriting was the more obvious behaviour anyways.

This is actually what happens already, but I'll be adding a test for it.

mozbhearsum commented on src/auslib/services/releases.py:
```diff
@@ -0,0 +1,379 @@
+            release = release[0]
+            if "scheduled_changes" not in release:
+                release["scheduled_changes"] = []
+        else:
```
Yep! Your comment makes me wonder if we should explicitly not support that though...

nthomas-mozilla replied:
I could go either way. If we can't think of any user confusion or malicious confusion scenarios then completeness is good.

mozbhearsum commented on src/auslib/services/releases.py:
```diff
@@ -0,0 +1,379 @@
+    # that it won't change in the future.
+    column = dbo.releases_json.data["schema_version"]
+    column.anon_label = "schema_version"
+    return dbo.releases_json.select(where={"name": name}, columns=[column], transaction=trans)[0]["schema_version"]
```
I can see why this is weird or confusing - it's syntactic sugar for doing deep queries into JSON data that result in SQL such as:
```
SELECT JSON_EXTRACT(releases_json.data, "$.schema_version") AS schema_version 
FROM releases_json
WHERE name = "Firefox-56.0-build1"
```

`data` is a [JSON](https://docs.sqlalchemy.org/en/13/core/type_basics.html#sqlalchemy.types.JSON) object here, which implements this sort of thing.

nthomas-mozilla replied:
Thanks for clearing that up.

mozbhearsum commented:
> Not sure what to suggest to improve on in the former - it just has to do a lot of manipulation - maybe some more docs on the helper functions there to make their inputs and outputs clearer, or some unit tests for them.

I'll see what I can do here. Was it just the helper functions that were tough to understand, or did you find the primary functions (`get_releases`, `update_release`, etc.) troublesome as well?

mozbhearsum commented on tests/admin/views/test_releases_v2.py:
```diff
@@ -0,0 +1,1173 @@
+                "scheduled_changes": [],
+            },
+            {
+                "name": "Firefox-64.0-build1",
```
I can do this, but it actually requires changing the impementation a bit (to sort the Releases before returning them rather than place them in the list in the order we find them). This is probably a sensible thing to do anyways.

nthomas-mozilla replied:
Oh well, if it's straight forward I guess. Not going to fight for this hill.

nthomas-mozilla commented:
> I'll see what I can do here. Was it just the helper functions that were tough to understand, or did you find the primary functions (`get_releases`, `update_release`, etc.) troublesome as well?

The helper functions mostly. There are different data structures coming out of them so an example would help with jumping around the code by eye, vs breaking out a debugger.

mozbhearsum commented on tests/admin/views/test_releases_v2.py:
```diff
@@ -0,0 +1,1173 @@
+    assert ret.status_code == 200, ret.data
+    assert ret.json == new_data_versions, ret.json
+
+    base_sc = dbo.releases_json.scheduled_changes.t.select().where(dbo.releases_json.scheduled_changes.sc_id == 18).execute().fetchall()
```
Whoops, that's a great point. Considering the API modifies by name it probably makes sense for us to be verifying by pulling by name, like you suggest.

mozbhearsum commented:
I think I've addressed all of the comments. I've left the existing commits untouched. Everything from 
e701c4b onwards is new code. Full diff vs. what you originally reviewed is at https://github.com/mozilla-releng/balrog/pull/1212/files/6d5add647a27c0f129733ab989721e4040d24383..0965bd023187ccdd39e7867b02fee48bf13a5dac (TIL that Github lets you do this in the "Files Changed" UI).

> > I'll see what I can do here. Was it just the helper functions that were tough to understand, or did you find the primary functions (`get_releases`, `update_release`, etc.) troublesome as well?

> 
> The helper functions mostly. There are different data structures coming out of them so an example would help with jumping around the code by eye, vs breaking out a debugger.

While looking at this I realized that they could be further simplified. Most notably - `separate_assets` doesn't require a dictionary with the base parts stripped away - it already simply ignores them. This means that `separate` (nee `do_split_release`) also doesn't have to deal with storing and returning assets. Along with that simplification I also improved some of the naming and docstrings.

As of yet, I haven't added any targeted tests for these functions specifically, but if you still think that would be useful I'm happy to do so. I'm not sure whether the existing improvements make things clear enough or not.

nthomas-mozilla commented on tests/admin/views/test_releases_v2.py:
```diff
@@ -0,0 +1,1293 @@
+
+    old_data_versions = populate_versions_dict(firefox_60_0b3_build1)
+    new_data_versions = versions_dict(default=2)
+    assert new_data_versions["platforms"]["Linux_x86_64-gcc3"]["locales"]["en-US"]
```
I feel uneasy about `new_data_versions` - a full dict of 2's when only a couple of entries are expected to take that value. The assert must be only checking keys in `ret.json`, otherwise it would fail.

To get closer to having an explicit expectation while holding onto the convenience of creating that, what about 
* using `versions_dict(default=None)`
* explicitly setting the expected values on the base and the modified locale

I'm hoping that will error out if any error creeps into the api response (ie 2 is not equal to None).

mozbhearsum replied:
> I feel uneasy about `new_data_versions` - a full dict of 2's when only a couple of entries are expected to take that value. The assert must be only checking keys in `ret.json`, otherwise it would fail.

I think I might be understanding your comment, but to be clear: I think `new_data_versions` only contains a `.` entry and one for `["platforms"]["Linux_x86_64-gcc3"]["locales"]["en-US"]` since `versions_dict` doesn't do any population aside from `.`.

> To get closer to having an explicit expectation while holding onto the convenience of creating that, what about
> 
>     * using `versions_dict(default=None)`
>     * explicitly setting the expected values on the base and the modified locale

Regardless of the above, I can see it being useful to avoid setting `.` at all in `versions_dict` and having the test methods do that.

nthomas-mozilla replied:
> > I feel uneasy about `new_data_versions` - a full dict of 2's when only a couple of entries are expected to take that value. The assert must be only checking keys in `ret.json`, otherwise it would fail.
> 
> I think I might be understanding your comment, but to be clear: I think `new_data_versions` only contains a `.` entry and one for `["platforms"]["Linux_x86_64-gcc3"]["locales"]["en-US"]` since `versions_dict` doesn't do any population aside from `.`.

I was thinking of the case where ret.json contains more data than it should and still passes the test, because of an incorrect assumption on my part, but it doesn't appear that this happens . A `versions_dict` will create a new key/default when it's accessed or asserted, but not when doing an equality comparison with `ret.json`. 

> Regardless of the above, I can see it being useful to avoid setting `.` at all in `versions_dict` and having the test methods do that.

Do you mean for clarity in the test rather than sorta-obscuring it with the include_base parameter on versions_dict ?

mozbhearsum replied:
> > > I feel uneasy about `new_data_versions` - a full dict of 2's when only a couple of entries are expected to take that value. The assert must be only checking keys in `ret.json`, otherwise it would fail.
> > 
> > 
> > I think I might be understanding your comment, but to be clear: I think `new_data_versions` only contains a `.` entry and one for `["platforms"]["Linux_x86_64-gcc3"]["locales"]["en-US"]` since `versions_dict` doesn't do any population aside from `.`.
> 
> I was thinking of the case where ret.json contains more data than it should and still passes the test, because of an incorrect assumption on my part, but it doesn't appear that this happens . A `versions_dict` will create a new key/default when it's accessed or asserted, but not when doing an equality comparison with `ret.json`.

Ahhhhh, I see. I'm actually a little surprised this is true now that you mention it. I'm guessing that the assert walks through the object with `.keys()`, which wouldn't trigger the defaults...

> 
> > Regardless of the above, I can see it being useful to avoid setting `.` at all in `versions_dict` and having the test methods do that.
> 
> Do you mean for clarity in the test rather than sorta-obscuring it with the include_base parameter on versions_dict ?

Yeah, exactly.

nthomas-mozilla commented on tests/admin/views/test_releases_v2.py:
```diff
@@ -0,0 +1,1293 @@
+def test_post_fails_when_release_doesnt_exist(api):
+    blob = {"detailsUrl": "https://newurl", "platforms": {"Darwin_x86_64-gcc3-u-i386-x86_64": {"locales": {"de": {"buildID": "999999999999999"}}}}}
+    old_data_versions = versions_dict()
+    assert old_data_versions["platforms"]["Darwin_x86_64-gcc3-u-i386-x86_64"]["locales"]["de"]
```
`populate_versions_dict(blob)` would be option here, given the test is about getting an error before trying to modify a release ?

mozbhearsum replied:
Could do, though that might be a less realistic request?

nthomas-mozilla replied:
Perhaps I'm missing something, but `old_data_versions = populate_versions_dict(blob)` seems to produce the same dict as the two lines here.

nthomas-mozilla replied:
On reading other tests (like a tweaked `test_put_add_scheduled_change_base_and_locale`) I see that the suggestion would be sending unnecessary data. Where`populate_versions_dict` is used it's for big blobs, although with a similar amount of change as this test. So I guess I don't feel hugely strongly any more, but something something consistency ?

mozbhearsum replied:
> Perhaps I'm missing something, but old_data_versions = populate_versions_dict(blob) seems to produce the same dict as the two lines here.

With fresh eyes, I think you're right about this. My brain was shortcutting `populate_versions_dict` as "always adds many platforms and locales". We may as well use `populate_versions_dict` for consistency, as you suggest.

nthomas-mozilla commented:
> I think I've addressed all of the comments. I've left the existing commits untouched. 

Thanks x2.

> As of yet, I haven't added any targeted tests for these functions specifically, but if you still think that would be useful I'm happy to do so. I'm not sure whether the existing improvements make things clear enough or not.

Given your recent comments about testing and contra-variance, which make sense to me too, I'm happy to leave the tests as they are.

mozbhearsum commented:
We talked yesterday and we agreed that everything here was OK now, but we want to add a test to make sure the `versions_dict` behaviour doesn't change in such a way that breaks all of our tests in mysterious ways. I'll be doing that before merging this.

mozbhearsum commented:
@nthomas-mozilla - I'm going to go ahead and merge this. If the canary test (or anything else) needs to be addressed just let me know and I'll follow up.
